### PR TITLE
Add elevation scale

### PR DESCRIPTION
**Pull request by jameskoster**
elevation

This PR adds variables for the elevation scale as outlined in https://github.com/WordPress/gutenberg/issues/63757. It follows the method applied in https://github.com/WordPress/gutenberg/pull/64007 to add the radius scale. 

* Adds new variables and guidelines
* Soft-deprecates old variables (`$shadow-popover` and `$shadow-modal`) by mapping them to the new ones

### Elevation scale

There are four values in the scale:

* `$elevation-x-small`: For sections and containers that group related content and controls, which may overlap other content. Example: Preview Frame.
* `$elevation-small`: For components that provide contextual feedback without being intrusive. Generally non-interruptive. Example: Tooltips, Snackbar.
* `$elevation-medium`: For components that offer additional actions. Example: Menus, Command Palette
* `$elevation-large`: For components that confirm decisions or handle necessary interruptions. Example: Modals.

In terms of values, I used the existing `$shadow-popover` for `$elevation-x-small`, and `$shadow-modal` for `$elevation-large`. `$elevation-small` and `$elevation-medium` are new values that I created by hand to slot between the existing values. It's possible these will change when it comes to applying the new scale across components. 

### Application
| `$elevation-x-small` | `$elevation-small` | `$elevation-medium` | `elevation-large` |
| --- | --- | --- | --- |
| Content / Preview frame<br>ZStack<br>Potential input elements such as; Range handle, Active item in ToggleGroupControl, Toggle handle<br>Resize handle<br>Block toolbar | Snackbar<br>Tooltip | DropdownMenu<br>Popover<br>Command Palette | Modal<br>ConfirmDialog<br>Guide |

### CODE REVIEW

**Comment by ciampo** [comment]
A few thoughts off the top of my head:

- Let's make sure we have clear instructions on how to translate the new scale to existing components like [`Elevation`](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-elevation--docs) and the components that consume it (like [`Card`](https://wordpress.github.io/gutenberg/?path=/docs/components-card--docs)) and [ZStack](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-zstack--docs)
- Let's review the `Popover` component and its usage
- Once we agree on the new proposed scale and its variable naming, let's agree on a list of follow-up tasks to clean up, lint, and deprecate the older variables like in https://github.com/WordPress/gutenberg/pull/64007
- How does the radii and elevation work relate to the `@wordpress/theme` effort?

**Comment by jameskoster** [comment]
> Let's make sure we have clear instructions on how to translate the new scale to existing components like [Elevation](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-elevation--docs) and the components that consume it (like [Card](https://wordpress.github.io/gutenberg/?path=/docs/components-card--docs)) and [ZStack](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-zstack--docs)

Would it make sense to simply map each step in the scale to the `value` prop in `Elevation`? Anything above 4 would just use the value associated with the fourth step.

> Let's review the Popover component and its usage

Are you asking which value would be applied to Popovers? Generally they contain un-interruptive configuration UI's, so `$elevation-30` probably makes most sense. It would be good to add well-defined guidelines around when to use popovers.

> How does the radii and elevation work relate to the @wordpress/theme effort?

That's a good question. In terms of elevation the shadow styles will need to be different for light / dark modes. Darker colored themes will require darker shadow styles in order for them to be visible. 

I don't _think_ radius should be affected, but another consideration is spacing; the theme component may need to support a 'density' option that affects spacing, type styles, components, etc. That part isn't really clear yet.

**Comment by ciampo** [comment]
> Would it make sense to simply map each step in the scale to the `value` prop in `Elevation`? Anything above 4 would just use the value associated with the fourth step.

That could work, although it could create unexpected changes in consumers. If we went down that way, we'd need to make sure to communicate it to consumers of those components via a dev note.

> Are you asking which value would be applied to Popovers? Generally they contain un-interruptive configuration UI's, so $elevation-30 probably makes most sense. It would be good to add well-defined guidelines around when to use popovers.

Absolutely. Changing the default popover box shadow is trivial, but what I really meant was that this is a great opportunity to review all popovers across Gutenberg and standardize their box shadow for better consistency and clarity.

**Comment by tyxla** [comment]
> Would it make sense to simply map each step in the scale to the value prop in Elevation? Anything above 4 would just use the value associated with the fourth step.

I think this would create discrepancies with the existing styles of `Elevation`. Perhaps we could provide a different syntax for the 4 scale values - for example, similar to the radius scale variable names in #64007?

**Comment by jasmussen** [comment]
Just capturing some small feedback/questions from [here](https://github.com/WordPress/gutenberg/issues/63757#issuecomment-2259813272), relevant to the PR.

**Comment by richtabor** [comment]
Is this more of a system for elevation, or proposing what those elevations should be, or both?

**Comment by jameskoster** [comment]
> Perhaps we could provide a different syntax for the 4 scale values - for example, similar to the radius scale variable names in https://github.com/WordPress/gutenberg/pull/64007?

That could work, although the proposed naming convention might lead to confusion? Would it be strange if `<Elevation value={ 1 } />` produced a stronger shadow than `<Elevation value={ 10 } />`? Or are you suggesting something different?

> Is this more of a system for elevation, or proposing what those elevations should be, or both?

The former, to avoid changing anything in production for now. We can define and apply values separately once we're happy with the scale.

@jasmussen If we want fewer steps in the scale I'd be tempted to combine `10` and `20`. We should probably work on some mockups (with rough shadow values) to help decide this.

**Comment by jasmussen** [comment]
> If we want fewer steps in the scale I'd be tempted to combine 10 and 20. We should probably work on some mockups (with rough shadow values) to help decide this.

Seems fine with me. 

I'm always just in the category of _start with as little as possible_ and then expand only when the need very clearly presents itself.

**Comment by hanneslsm** [comment]
I don't think it's currently needed because it doesn't fit the common logic of elevation (Level 0 and up) and the current design style (would be more skeuomorphic) but want to bring it up anyways, just to be sure:
negative elevation / inset shadow / depth. That'd influence the naming.
Examples could be iFrame or input fields or even buttons.

**Comment by tyxla** [comment]
> > Perhaps we could provide a different syntax for the 4 scale values - for example, similar to the radius scale variable names in https://github.com/WordPress/gutenberg/pull/64007?
> 
> That could work, although the proposed naming convention might lead to confusion? Would it be strange if <Elevation value={ 1 } /> produced a stronger shadow than <Elevation value={ 10 } />? Or are you suggesting something different?

My thoughts were around using string `small`, `medum`, `large`, etc. for those 4 particular sizes, and those would convert to their corresponding `Elevation` values. In addition, `Elevation` would continue supporting number values as before.

**Comment by jameskoster** [comment]
@hanneslsm shadow styles are already being applied albeit sporadically without well defined intent. 

A big part of this effort is about defining how/when to use elevation as we embark on the broader admin design effort (https://github.com/WordPress/gutenberg/issues/53322). Please refer to the four suggested steps in the scale which details this. 

In terms of mockups it's a little bit tricky because there are as yet undecided details to consider. For instance details like radius, color schemes, type, light/dark modes etc. With that said consider the following rough in terms of those details (including the shadows); they're primarily to demonstrate the 4 steps in the scale in some existing flows and experiences.

### `$elevation-10` applied to content frame.

Elevation 1

### `$elevation-20` applied to tooltip, snackbar, bulk-actions toolbar.
Elevation 2

### `$elevation-30` applied to menu.
Elevation 3

### `$elevation-40` applied to modal.
Elevation 4

---

> My thoughts were around using string small, medum, large, etc. 

@tyxla it can work, but leaves less room for manoeuvrability if we later need to add steps in the scale.

**Comment by hanneslsm** [comment]
> @hanneslsm shadow styles are already being applied albeit sporadically without well defined intent.

> A big part of this effort is about defining how/when to use elevation as we embark on the broader admin design effort (https://github.com/WordPress/gutenberg/issues/53322). Please refer to the four suggested steps in the scale which details this.

Thanks for clarifying! Looks good!
In terms of naming I'm also in favor of the numeric scale by the way.

**Comment by jameskoster** [comment]
I don't think there's a huge amount in it, but I've switched the names to match the radius variables, IE to use `-small`, `-medium` suffixes. Reason being:

* It's more human readable.
* Maps more cleanly to the existing `Elevation` component.
* It seems unlikely we'll need to add more steps in the future, but if necessary values like `$elevation-x-large` can be added, and components remapped as required.
* Consistency with radius

I also added some examples of where the different values would be applied across the existing set of components. 

`/packages/components/src/utils/config-values.js` is also updated.

**Review by DaniGuardiola — APPROVED** [review]
This looks correct to me from a technical standpoint. The only concern I have is that the naming might be confusing since it doesn't contain "shadow", yet it's supposed to be an option for shadow property values. However, it's not a big deal so not going to block on that.

Appreciate the comments with added context on usage of each option, it's a nice plus.

**Review by DaniGuardiola — CHANGES_REQUESTED** [review]
Ah, missed that this is currently broken because `$shadow-popover` is assigned `$elevation-10` (similar story with "modal"), guess it's just out of sync after recent changes in naming. After this is corrected, my previous (approval) review stands. Example failure in CI: https://github.com/WordPress/gutenberg/actions/runs/10249250569/job/28352231809?pr=64108

image

**Comment by jameskoster** [comment]
Good spot, I fixed that.

**Comment by jameskoster** [comment]
@DaniGuardiola @ciampo do you think this needs a components changelog entry?

**Comment by DaniGuardiola** [comment]
@jameskoster good question. @ciampo and I discussed and we both think it's not necessary.

**Review by DaniGuardiola — APPROVED** [review]
Issue was addressed 👍

File: packages/components/src/utils/config-values.js
Comment (ciampo) [inline packages/components/src/utils/config-values.js:76]: Since I don't think that `popoverShadow`  is exported from the package, I think it should be OK to remove it and replace its usages with `elevationXSmall`.

This can be definitely done in one of the follow-up PRs

**Comment by jameskoster** [comment]
Cool. I'll go ahead and merge this. We can continue to refine the scale and values as needed when we begin applying across components.

**Pull request merged**
